diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index 0a18a43544..9063565159 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -2342,6 +2342,9 @@ int OPENSSL_init_ssl(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings);
 __owur const struct openssl_ssl_test_functions *SSL_test_functions(void);
 # endif
 
+__owur unsigned char * SSL_RECORD_LAYER_get_packet(SSL *ssl);
+__owur size_t SSL_RECORD_LAYER_get_packet_length(SSL *ssl);
+
 __owur int SSL_free_buffers(SSL *ssl);
 __owur int SSL_alloc_buffers(SSL *ssl);
 
diff --git a/include/openssl/tls1.h b/include/openssl/tls1.h
index 2e46cf80d3..b19719fc21 100644
--- a/include/openssl/tls1.h
+++ b/include/openssl/tls1.h
@@ -131,6 +131,11 @@ extern "C" {
 /* ExtensionType value from RFC7627 */
 # define TLSEXT_TYPE_extended_master_secret      23
 
+/* [draft-ietf-tls-certificate-compression] */
+# define TLSEXT_TYPE_compress_certificate        27
+/* ExtensionType value from RFC8449 */
+# define TLSEXT_TYPE_record_size_limit           28
+
 /* ExtensionType value from RFC4507 */
 # define TLSEXT_TYPE_session_ticket              35
 
diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index d75158e30c..56747533c8 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -5063,6 +5063,19 @@ int SSL_client_hello_get0_ext(SSL *s, unsigned int type, const unsigned char **o
     return 0;
 }
 
+unsigned char * SSL_RECORD_LAYER_get_packet(SSL *ssl)
+{
+    RECORD_LAYER *rl = &ssl->rlayer;
+    return RECORD_LAYER_get_packet(rl);
+}
+
+
+size_t SSL_RECORD_LAYER_get_packet_length(SSL *ssl)
+{
+    RECORD_LAYER *rl = &ssl->rlayer;
+    return RECORD_LAYER_get_packet_length(rl);
+}
+
 int SSL_free_buffers(SSL *ssl)
 {
     RECORD_LAYER *rl = &ssl->rlayer;
diff --git a/ssl/ssl_locl.h b/ssl/ssl_locl.h
index e8819e7a28..40cc101afc 100644
--- a/ssl/ssl_locl.h
+++ b/ssl/ssl_locl.h
@@ -719,6 +719,8 @@ typedef enum tlsext_index_en {
     TLSEXT_IDX_cryptopro_bug,
     TLSEXT_IDX_early_data,
     TLSEXT_IDX_certificate_authorities,
+    TLSEXT_IDX_compress_certificate,
+    TLSEXT_IDX_record_size_limit,
     TLSEXT_IDX_padding,
     TLSEXT_IDX_psk,
     /* Dummy index - must always be the last entry */
diff --git a/ssl/statem/extensions.c b/ssl/statem/extensions.c
index 8422161dc1..2bb33205f1 100644
--- a/ssl/statem/extensions.c
+++ b/ssl/statem/extensions.c
@@ -371,6 +371,22 @@ static const EXTENSION_DEFINITION ext_defs[] = {
         tls_construct_certificate_authorities,
         tls_construct_certificate_authorities, NULL,
     },
+    {
+        TLSEXT_TYPE_compress_certificate,
+        SSL_EXT_CLIENT_HELLO,
+        NULL,
+        NULL, NULL,
+        NULL,
+        NULL, NULL,
+    },
+    {
+        TLSEXT_TYPE_record_size_limit,
+        SSL_EXT_CLIENT_HELLO,
+        NULL,
+        NULL, NULL,
+        NULL,
+        NULL, NULL,
+    },
     {
         /* Must be immediately before pre_shared_key */
         TLSEXT_TYPE_padding,
diff --git a/util/libssl.num b/util/libssl.num
index ccf73411a9..1224c5ee71 100644
--- a/util/libssl.num
+++ b/util/libssl.num
@@ -497,3 +497,5 @@ SSL_get_recv_max_early_data             497	1_1_1	EXIST::FUNCTION:
 SSL_CTX_get_recv_max_early_data         498	1_1_1	EXIST::FUNCTION:
 SSL_CTX_set_recv_max_early_data         499	1_1_1	EXIST::FUNCTION:
 SSL_CTX_set_post_handshake_auth         500	1_1_1	EXIST::FUNCTION:
+SSL_RECORD_LAYER_get_packet             501	1_1_1	EXIST::FUNCTION:
+SSL_RECORD_LAYER_get_packet_length      502	1_1_1	EXIST::FUNCTION:
